File: python/replace-words.py
from typing import List


class TrieNode:
    def __init__(self):
        self.children = {}
        self.last = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, str) -> None:
        curr = self.root
        for c in str:
            if c not in curr.children:
                curr.children[c] = TrieNode()
            curr = curr.children[c]
        curr.last = True

    def find(self, str) -> int:
        root = ""
        curr = self.root
        for i, c in enumerate(str):
            if c in curr.children:
                root += c
                curr = curr.children[c]
                if curr.last == True:
                    return (True, root)
            else:
                return (False, "")
        return (True, root)


class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        book = Trie()
        for word in dictionary:
            book.insert(word)
        arr = sentence.split(" ")
        for i, deriv in enumerate(arr):
            check = book.find(deriv)
            if check[0]:
                arr[i] = check[1]
        return " ".join(arr)

File: typescript/medium/replace-words.ts
class Trie {
    map: Map<string, Trie>;
    end: string | boolean;

    constructor() {
        this.map = new Map();
        this.end = false;
    }
}

function replaceWords(dictionary: string[], sentence: string): string {
    const trie = new Trie();
    for (const root of dictionary) {
        let it = trie;
        for (const char of root) {
            if (!it.map.has(char)) it.map.set(char, new Trie());
            it = it.map.get(char)!;
        }
        it.end = root;
    }

    const words = sentence.split(' ');
    let res: string[] = [];
    for (const word of words) {
        res.push(word);
        let it = trie;
        for (const char of word) {
            if (!it.map.has(char)) break;
            it = it.map.get(char)!;
            if (it.end !== false) {
                res.pop();
                res.push(it.end as string);
                break;
            }
        }
    }
    return res.join(' ');
}